### PR TITLE
Linear autodiff reland

### PR DESCRIPTION
**Pull request by jjsjann123**
add backward for aten::linear to SymbolicScript to make aten::linear differentiable in TorchScript
make torch.nn.functional.linear call aten::linear to avoid python control flow for that op
onnx changes that are needed to support this

reland of  PR #50856

### CODE REVIEW

**Comment by jjsjann123** [comment]
I'm still building to test locally.
patched sent by @XiaobingSuper looks good to me so I'm kicking off CI as well.

**Review by XiaobingSuper — APPROVED** [review]

**Comment by jjsjann123** [comment]
Updated BW compatibility list to allow argument name change.
Previous cuda IMA failure can't be repro'ed locally (and doesn't look like it's related at first glance). I'm trying my luck with merging master.

File: test/backward_compatibility/check_backward_compatibility.py [hunk 1]
```diff
@@ -71,6 +71,7 @@
     ("aten::_foreach_div_", datetime.date(2021, 3, 2)),
     ("aten::_foreach_addcdiv", datetime.date(2021, 3, 2)),
     ("aten::_foreach_addcmul", datetime.date(2021, 3, 2)),
+    ("aten::mkl_linear", datetime.date(2021, 3, 2)),
```
Comment (XiaobingSuper) [inline test/backward_compatibility/check_backward_compatibility.py:74]: it should be  **mkldnn_linear**.

Comment (jjsjann123) [reply test/backward_compatibility/check_backward_compatibility.py:74]: Oops, Will update after all tests finished (just to rule out the failure we had earlier).

**Comment by jjsjann123** [comment]
Trying again since the foul PR has been reverted.

File: tools/autograd/derivatives.yaml [hunk 2]
```diff
@@ -187,9 +187,9 @@
   tensor2: handle_r_to_c(tensor2.scalar_type(), grad * (tensor1 * value).conj())
 
 - name: addmm(Tensor self, Tensor mat1, Tensor mat2, *, Scalar beta=1, Scalar alpha=1) -> Tensor
-  self: "grad.is_mkldnn() ? grad.to_dense() : maybe_multiply(grad, beta.conj())"
-  mat1: "grad.is_mkldnn() ? mkldnn_linear_backward_input(mat1.sizes(), grad, mat2.t()) : mm_mat1_backward(grad, mat2, mat1.sizes(), mat1.strides(), alpha)"
-  mat2: "grad.is_mkldnn() ? (std::get<0>(mkldnn_linear_backward_weights(grad, mat1, mat2.t(), true))).t() : mm_mat2_backward(grad, mat1, mat2.sizes(), mat2.strides(), alpha)"
+  self: maybe_multiply(grad, beta.conj())
```
Comment (ngimel) [inline tools/autograd/derivatives.yaml:190]: wait, how is addmm for mkldnn tensors supposed to work now? These lines were introduced by #49453, and they are needed. 
@XiaobingSuper why are there no tests that would catch this?

Comment (jjsjann123) [reply tools/autograd/derivatives.yaml:190]: I thought the other PR is trying to enable it for linear, and has to go through admm since it is how linear was represented in JIT.
So that part is done via autograd below:
```
- name: mkldnn_linear(Tensor self, Tensor weight, Tensor? bias=None) -> Tensor
  self, weight, bias: mkldnn_linear_backward(self, grad, weight, grad_input_mask)
```

I don't think the admm part would be trouble some in our context. So we could try to put them back if it's intended to be kept.

Comment (ngimel) [reply tools/autograd/derivatives.yaml:190]: I don't see this PR removing `mkldnn_addmm_wraper` from native_functions.yaml, which means that addmm backward is needed.

Comment (ngimel) [reply tools/autograd/derivatives.yaml:190]: Ah, sorry, it's actually removed, fine then.

File: torch/csrc/jit/runtime/symbolic_script.cpp [hunk 3]
```diff
@@ -403,6 +403,23 @@ const std::vector<std::string> functions = {
                 return grad_self, grad_other
 
             return torch.matmul(self, other), backward
+
+        def linear(input : Tensor,
+                   weight : Tensor,
+                   bias : Optional[Tensor]):
+            result = torch.linear(input, weight, bias)
```
Comment (ngimel) [inline torch/csrc/jit/runtime/symbolic_script.cpp:410]: what's torch.linear and how does this work?

Comment (jjsjann123) [reply torch/csrc/jit/runtime/symbolic_script.cpp:410]: This goes through:
https://github.com/pytorch/pytorch/blob/00675292ca52aa459690b9a31f5893f53dcf8659/aten/src/ATen/native/native_functions.yaml#L2133-L2135
https://github.com/pytorch/pytorch/blob/00675292ca52aa459690b9a31f5893f53dcf8659/aten/src/ATen/native/Linear.cpp#L30-L48

File: tools/autograd/derivatives.yaml (hunk 2 above)
Comment (jjsjann123) [reply tools/autograd/derivatives.yaml:190]: Let me try to throw a test there and patch it.

File: torch/csrc/jit/runtime/symbolic_script.cpp (hunk 3 above)
Comment (ngimel) [reply torch/csrc/jit/runtime/symbolic_script.cpp:410]: Nope, because native_functions entry has `python_module: nn`, it's not available as torch.linear (try it), it's only available as torch._C._nn.linear (that's what `linear` in torch/nn/functiona.py calls).

File: tools/autograd/derivatives.yaml (hunk 2 above)
Comment (ngimel) [reply tools/autograd/derivatives.yaml:190]: No, please don't bring mkldnn_addmm_wraper back, it's good that it's removed, without your PR addmm for mkldnn tensors was in pretty bad shape, but now luckily we don't need to care about it.

Comment (jjsjann123) [reply tools/autograd/derivatives.yaml:190]: I notice that we did indeed remove `mkldnn_addmm_wraper` and from the comment it's intended

https://github.com/pytorch/pytorch/pull/51613/files/919a29a171b6c26bc70e8b88511bfbe148d5d978#diff-172c536bf0e2a56a98b3b07941391e3630fff843b5707a606f1fb5eafac2c35cL16

I'm trying a build to revert the patch and only apply the autograd rule for `mkldnn_linear` (so we can keep mkldnn_addmm). Please let me know if this is needed. @XiaobingSuper

Comment (ngimel) [reply tools/autograd/derivatives.yaml:190]: No, we don't want to keep mkldnn_addmm

Comment (jjsjann123) [reply tools/autograd/derivatives.yaml:190]: Got it.

I actually got my local build to work and `mkldnn_addmm_wraper` is giving me wrong results.
I traced it down to aten::mkldnn_linear. it can't handle non-contiguous weight. Since we are not using the wrapper which does that transpose, it's not an issue any more. CC'ing @XiaobingSuper for future reference.

Comment (XiaobingSuper) [reply tools/autograd/derivatives.yaml:190]: **mkldnn_addmm_wraper** is for call in mkldnn path for some linear case(input' s dim is two, and has bias, linear will call in addmm), if  aten linear ported, we can removed it.

File: torch/csrc/jit/runtime/symbolic_script.cpp (hunk 3 above)
Comment (jjsjann123) [reply torch/csrc/jit/runtime/symbolic_script.cpp:410]: strings in symbolic_script.cpp is parsed directly, so it has nothing to do with how aten is exposed in python.
specifically here, we are using compilation unit to parse the string directly and `torch` is resolved by NativeResolver where they map it to `aten`, so it is mapped to `aten::linear`: https://github.com/pytorch/pytorch/blob/fc1153a8be7cb1d4e7d1219746dc744cb8e28b4f/torch/csrc/jit/frontend/resolver.h#L48-L56

Comment (jjsjann123) [reply torch/csrc/jit/runtime/symbolic_script.cpp:410]: At least that's how I understand this... Granted that I know very little about jit frontend so I'm cc'ing @Krovatkin here just in case I'm embarrassing myself again :crossed_fingers:

File: tools/autograd/derivatives.yaml (hunk 2 above)
Comment (XiaobingSuper) [reply tools/autograd/derivatives.yaml:190]: > wait, how is addmm for mkldnn tensors supposed to work now? These lines were introduced by #49453, and they are needed.
> @XiaobingSuper why are there no tests that would catch this?

#49453 is for catch linear case when linear call in addmm , there just has test linear test case.

File: torch/csrc/jit/runtime/symbolic_script.cpp (hunk 3 above)
Comment (ngimel) [reply torch/csrc/jit/runtime/symbolic_script.cpp:410]: THis is super confusing, but plausible

File: tools/autograd/derivatives.yaml (hunk 2 above)
Comment (jjsjann123) [reply tools/autograd/derivatives.yaml:190]: FYI #51713

**Review by ngimel — APPROVED** [review]

**Comment by jjsjann123** [comment]
Anything I can help with internal build error?
